src/Passes/Source/ProfilePass/Profile.cpp

```diff
@@ -0,0 +1,680 @@
+                            {
+                                queue.push_back(bb);
+                                blocks_queued.insert(bb);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Deleting constants
+        for (auto& x : schedule_instruction_deletion)
+        {
+            x->eraseFromParent();
+        }
+
+        --depth_;
+
+        return true;
+    }
+
+    bool ProfilePass::isActive(llvm::Value* value) const
+    {
+        return active_pieces_.find(value) != active_pieces_.end();
+    }
+
+    void ProfilePass::runCopyAndExpand(llvm::Module& module, llvm::ModuleAnalysisManager&)
+    {
+        replacements_.clear();
+        // For every instruction in every block, we attempt a match
+        // and replace.
+        for (auto& function : module)
+        {
+            if (function.hasFnAttribute("EntryPoint"))
+            {
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return copyAndExpand(value, modifier);
+                });
+            }
+        }
+
+        // Dead code detection
+        for (auto& function : module)
+        {
+            if (function.hasFnAttribute("EntryPoint"))
+            {
+                // Marking function as active
+                active_pieces_.insert(&function);
+
+                // Detectecting active code
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return detectActiveCode(value, modifier);
+                });
+            }
+        }
+
+        applyReplacements();
+    }
+
+    void ProfilePass::applyReplacements()
+    {
+        // Applying all replacements
+
+        std::unordered_set<llvm::Value*> already_removed;
+        for (auto it = replacements_.rbegin(); it != replacements_.rend(); ++it)
+        {
+            auto instr1 = llvm::dyn_cast<llvm::Instruction>(it->first);
+
+            // Checking if by accident the same instruction was added
+            if (already_removed.find(instr1) != already_removed.end())
+            {
+                throw std::runtime_error("Instruction was already removed.");
+            }
+            already_removed.insert(instr1);
+
+            if (instr1 == nullptr)
+            {
+                llvm::errs() << "; WARNING: cannot deal with non-instruction replacements\n";
+                continue;
+            }
+
+            // Cheking if have a replacement for the instruction
+            if (it->second != nullptr)
+            {
+                // ... if so, we just replace it,
+                auto instr2 = llvm::dyn_cast<llvm::Instruction>(it->second);
+                if (instr2 == nullptr)
+                {
+                    llvm::errs() << "; WARNING: cannot replace instruction with non-instruction\n";
```

Comment:
When would this case occur?

Reply:
Again, this may never occur and just like the above, it should be reported as an internal error.